README.txt
Perl-Ogre
=========

This is a Perl binding for OGRE, Object-Oriented Graphics Rendering Engine,
a C++ library found at http://www.ogre3d.org/ .

The wrapping is currently incomplete, but there are several interesting
examples working now (see examples/README.txt).
See TODO.txt for the things I'd like to do. Suggest more things
(preferably submitted with patches) if you like.


DEPENDENCIES

Only versions >= 1.4.3 (current latest release as of August 2007) of
OGRE are currently supported (it's the version I have installed),
so you need to install that either by building from source or by installing
a package. If you use Ubuntu, I have some notes below that might help.

Later 1.4.x versions will probably work also. (As of 15 Dec 2007, I'm using
version 1.4.5 on Ubuntu and haven't noticed any problems.)

Makefile.PL uses pkg-config to get information about the libraries and header
files needed to build against OGRE, so you should be able to do this:

  pkg-config --libs OGRE
  pkg-config --cflags OGRE
  pkg-config --modversion OGRE

The last one should say at least 1.4.3.

The C++ compiler used by default is `g++`, but you can specify a different
C++ compiler by setting the CXX environmental variable. Anything more,
and you'll have to hack at Makefile.PL.

I have the impression that OGRE doesn't install pkg-config on Windows
(though nobody ever answers when I ask...), so this package will probably
not currently work on Windows. I'd like it to, but I don't use Windows.
Please let me know how if you get it working.


OPTIONAL

If you have Gtk2 installed, which in practice means that this is true:

  pkg-config --atleast-version=2.0.0 gtk+-2.0

then a static method, Ogre->getWindowHandleString, will be built.
The string returned can be passed to the `params' argument of
Ogre::Root::createRenderWindow in order to use a window (widget)
already created by gtk2-perl or wxPerl, thereby letting you embed
an Ogre widget in a GUI application. This feature is currently
EXPERIMENTAL, meaning its usage might change in the future
or be dropped altogether if it doesn't work.

Some scripts under examples/ require these Perl modules:

 - OIS
 - Readonly

I recommend installing both of them, but you don't have to. 

Note for OIS: there's apparently an incompatibility between versions
0.99_rc1 (which I have installed) and 1.0.0 of libois; in particular
in OIS/xs/InputManager.xs you might get two methods failing to work:
s/numJoysticks/numJoySticks/ and s/numKeyBoards/numKeyboards/.
I'll try to get that worked around soon.

It would be a good idea to make sure you also have installed CEGUI, OIS,
libdevil, and nVidia Cg, if those aren't already installed. This Perl module
currently only provides support for OGRE and OIS, however; I might wrap
CEGUI some day, the goal being to run OGRE's samples and tutorials from
Perl. (The tutorials and documentation for OGRE are excellent.
I highly recommend going through them before trying this Perl wrapper.)
The reason I recommend installing those libraries, is mainly just so
you can run the (C++) tutorials.


INSTALLATION

To install this module, do the usual:

   perl Makefile.PL
   make
   make test
   make install

You might have to edit Makefile.PL to get it to work for your system.
If so, please let me know.


INSTALLING OGRE UNDER UBUNTU

As of Feisty Fawn, there is a package for OGRE 1.4.3,
so that can be easily installed.

  # apt-get update

  # apt-get install libogre-dev libogre14
  # apt-get install nvidia-cg-toolkit 

  (you'll want these too, if they're not already installed)
  # apt-get install libcegui-mk2-doc libcegui-mk2-dev libcegui-mk2-1
  # apt-get install libois1 libois-dev
  # apt-get install libdevil-dev libdevil1c2

Note: these package names are as of when I installed them.
The names might change.


You can also create your own package by building from source.
Check out: http://www.ogre3d.org/wiki/index.php/FromSourceUbuntu
Seems like it should be pretty easy using `make checkinstall`.
Good luck with that.


COPYRIGHT AND LICENCE

Please report any bugs/suggestions to <slanning@cpan.org>.

Copyright 2007, Scott Lanning. All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

Beware that OGRE itself is under an LGPL license. See http://www.ogre3d.org/
for more (and probably more accurate) information.


Changes
0.35 2007-12-15 | Ogre::AL (new module)
- new (separate) CPAN module Ogre::AL, which is a binding for
  the OpenAL audio library and lets you attach sounds to scene nodes
- added three Perl-specific methods to OverlayManager, to account for
  C++ casting (see note for 0.33): createTextAreaOverlayElement,
  createPanelOverlayElement, createBorderPanelOverlayElement,
  which are all variations of createOverlayElement; also wrapped
  those three classes while I was at it
- added -Wno-write-strings to Makefile.PL if you compile with
  g++ >= 4.2, otherwise you would get literally a thousand warnings
  which I have no control over (I think these might be gone
  in perl 5.10, so I also check for that)
- bumped up OIS requirement in Ogre::ExampleListener to version 0.04
- added more operator overloads to Degree and Radian
- wrapped Ogre::MaterialManager to fix Ogre::BetaGUI

0.34 2007-12-09 | scene query enhancements
- wrapped SceneQuery, ManualObject, and AxisAlignedBox methods
  (for Ogre::BetaGUI itute3.pl and itute4.pl examples)
  note: for ManualObject, still not the constructor/destructor,
  though it can be created through SceneManager
- wrapped createPlaneBoundedVolumeQuery in SceneManager,
  which takes an array ref of PlaneBoundedVolume* instead of
  a PlaneBoundedVolumeList (which is a std::vector)
- added more support for RegionSceneQuery and its subclasses

0.33 2007-12-04 | Ogre::BetaGUI (new module)
- new (separate) CPAN module Ogre::BetaGUI; see it and its itute2.pl example
  demonstrating a mouse cursor
- added createOverlayContainer method to OverlayManager.xs;
  this Perl-specific method is equivalent to createOverlayElement
  but returns an OverlayContainer (static_cast) instead
- fixed Quaternion operator overloading
- added Vector2 constructor/destructor, tests

0.32 2007-11-26 | lighting.pl
- added lighting.pl example
- implemented Controller<Real>, ControllerValue<Real>, ControllerFunction<Real>
  for use with lighting example; note that Controller<Real> is called ControllerReal
  in the Perl binding
- fixed a little bug with setDiffuseColour/setSpecularColour of Ogre::Light
- (internal) renamed PerlOGREListener.h to PerlOGRECallback.h, etc. (generalized from
  listener to any callback class in Perl, in particular to handle
  ControllerValue and ControllerFunction)

0.31 2007-11-17 | terrain.pl
- added terrain.pl example
- wrapped RaySceneQuery class, among others
- updated docs (note: I need to update the gendocs.pl script,
  so there are some weird things like "$bool enabled" instead
  of "bool $enabled")

0.30 2007-11-16 | skeletalanim.pl
- added skeletalanim.pl example
- wrapped more classes, again to get the example working
- improved Vector3 and Quaternion operator overloading,
  so now the robot rotates in animate.pl instead of moonwalking
- fixed a few places where I'd neglected to put the return value
  of some methods into RETVAL, so segfaults would happen
- added init-class.pl and strip-api-docs.pl genscripts
  that I use for generating/initializing new classes

0.29 2007-11-14 | cameratrack.pl, particleFX.pl
- added cameratrack.pl and particleFX.pl examples
- wrapped many classes, mainly to make the examples work
- fixed a tiny bug in Ogre::ExampleFrameListener
- note: one reason there was a 2-month gap between releases
  was that I tried to write a script to generate the entire API;
  the aftermath of that attempt is genscripts/genxsubs.pl,
  but I've given up on it for now

0.28 2007-09-16 | gtk2-perl and wxPerl
- added a VERY EXPERIMENTAL feature to allow Ogre's render window to be
  an existing Gtk2 or Wx window instead of the defaultly generated one.
  See README.txt for details (you have to have gtk2 libs/headers installed).
  I kind of got it working for Gtk2, see examples/gtk2robot.pl.
  I couldn't get examples/wx.pl to actually work yet, though it's no longer
  segfaulting at least. (I think the Ogre window is underneath the other ones,
  i.e it's the "parent" window.)
- wrapped more methods in RenderWindow, Viewport

0.27 2007-09-12 | constants
- API CHANGE: fixed constants, adding all enums (and genconsts.pl script);
  now you export them normally, not call as static methods.
  Some export tags changed, too; they're all named after the corresponding
  enum's name now. (Note: OIS constants are still "broken".)
  Thanks to Vladimir Olenin for pointing me to `gccxml`, which was used
  in parsing the C++ headers.

0.26 2007-09-09 | documentation
- added some documentation (and gendocs.pl script which generates it)
- added DegRad typemap to handle both Degree or Radian types,
  so now Degree and Radian are interchangeable like in the C++ API
  (and wrapping those methods is simpler :)

0.25 2007-09-07 | animate.pl
- added animate.pl example, which demos animation
  (note: this is still a little incomplete, so the robot moonwalks
   part of the way)

0.24 2007-09-06 | minor improvements
- added typemap for Real and changed all occurrences of float to Real
  (which can be float or double)
- added constructors, overloaded (bool) operators, tests to:
  ColourValue, ConfigFile, Degree, Plane, Quaternion, Radian, Vector3

0.23 2007-09-04 | sky.pl
- moved ExampleFrameListener and ExampleApplication packages from
  examples/buffered.pl out under Ogre/ , so they can be used by other
  applications
- added sky.pl example, which implements OGRE's "Basic Tutorial 3"
- fixed unbuffered input in the Example* classes
- wrapped a lot of SceneManager

0.22 2007-09-02 | buffered.pl
- added buffered.pl, which implements OGRE's "Basic Tutorial 5"

0.21 2007-09-01 |
- added WindowEventListener functionality, though not "tested" yet
  (working on another example)

0.20 2007-08-30 | ninja.pl, listeningninja.pl
- added ninja.pl example, along with necessary wrappers
  to get it to work (you also need to install OIS,
  which I just uploaded to CPAN)
- also added listeningninja.pl example, along with
  FrameListener functionality

Makefile.PL
use Config;
use ExtUtils::MakeMaker;
use strict;
use warnings;

main();
exit 0;

sub main {
    my $fixed_args = fixed_args('Ogre');
    my $varied_args = varied_args();
    my $gpp_warn_args = gpp_warn_args();

    WriteMakefile(%$fixed_args, %$varied_args, %$gpp_warn_args);
}

sub gpp_warn_args {
    # if you're using gcc >= 4.2 and perl < 5.10, you'll probably
    # get warnings like this during `make` :
    #  OgreAL.c:83: warning: deprecated conversion from string constant to 'char*'
    # The C code generated by `xsubpp` uses literal strings (string constants)
    # as args to functions expecting char*, in particular newXSproto.
    # This disables those warnings.

    if ($Config{PERL_REVISION} == 5 && $Config{PERL_VERSION} < 10) {
        if (my_compiler() eq 'g++') {
            my $str = `g++ -dumpversion`;
            unless ($?) {
                chomp $str;
                my ($v, $s) = split /\./, $str;
                if (($v == 4 && $s >= 2) || $v > 4) {
                    return {'CCFLAGS' => '-Wno-write-strings'};
                }
            }
        }
    }

    # there will be no warnings, or we'll just tolerate them
    return {};
}

sub my_compiler {
    return $ENV{'CXX'} || 'g++';
}

sub fixed_args {
    my ($pkg) = @_;

    return {
        'NAME'          => $pkg,
        'VERSION_FROM'  => "$pkg.pm",
        'ABSTRACT_FROM' => "$pkg.pm",
        'AUTHOR'        => 'Scott Lanning <slanning@cpan.org>',
        'LD'            => '$(CC)',
        'OBJECT'        => '$(O_FILES)',
        'MAN3PODS'      => {},
        'XSOPT'         => '-C++',
        'TYPEMAPS'      => ['perlobject.map'],
    };
}

sub varied_args {
    # Make sure OGRE libs are known by pkg-config
    my $pkgname = 'OGRE';

    my @errors = ();
    push @errors, check_pkg_config($pkgname, '1.4.3');
    if (@errors) {
        die(map { "$_$/" } @errors);
    }

    # Get include dirs and defines
    my @cflags = ();
    push @cflags, pkg_config($pkgname, 'cflags');

    # Get lib dirs
    my @libs = ();
    push @libs, pkg_config($pkgname, 'libs');

    my @defines = ();


    # Check if Gtk+ is installed
    # (don't know what version is actually required, gtk2-perl uses 2.0.0)
    $pkgname = 'gtk+-2.0';

    @errors = check_pkg_config($pkgname, '2.0.0');
    if (@errors) {
        print "\nNote: gtk+ support not enabled.\nReasons:\n",
          map({ "- $_$/" } @errors);
        print "See README.txt for information on enabling Gtk2 support.\n";
    }
    else {
        print "\nEnabling gtk+ support.\n";
        push @cflags, pkg_config($pkgname, 'cflags');
        push @libs, pkg_config($pkgname, 'libs');
        push @defines, '-DPERLOGRE_HAS_GTK2';
    }


    my %prereqs = (
        'Test::More' => 0,
    );

    # Check if optional Perl modules are installed.
    my $OIS_REQ_VERSION = 0.04;
    unless (eval { require OIS && $OIS::VERSION >= $OIS_REQ_VERSION }) {
        my $msg = "\nNote: the Perl module OIS >= $OIS_REQ_VERSION is not installed,\n"
          . "so you won't be able to run some examples\n"
          . "or use Ogre::ExampleFrameListener.\n"
          . "Installing OIS would be a very good idea\n"
          . "unless you have some other way to handle keyboard and mouse input.\n"
          . "It's fine to install it after installing Ogre.\n\n"
          . "Do you want to install OIS now?";

        my $val = ExtUtils::MakeMaker::prompt($msg, 'n');
        if ($val =~ /^y/i) {
            $prereqs{'OIS'} = $OIS_REQ_VERSION;
            print "\nOIS >= $OIS_REQ_VERSION added to prerequisites.\n";
        }
    }
    unless (eval { require Readonly }) {
        my $msg = "\nNote: the Perl module Readonly is not installed,\n"
          . "so you won't be able to run some examples.\n"
          . "I recommend installing this module (and Readonly::XS if possible).\n"
          . "It's fine to install it after installing Ogre.\n\n"
          . "Do you want to install Readonly now?";
        my $val = ExtUtils::MakeMaker::prompt($msg, 'n');
        if ($val =~ /^y/i) {
            $prereqs{'Readonly'} = '0';
            print "\nReadonly added to prerequisites.\n";
        }
    }

    return {
        'PREREQ_PM'         => \%prereqs,
        'CC'                => my_compiler(),
        'INC'               => join(' ', @cflags),
        'LIBS'              => join(' ', @libs),
        (@defines ? ('DEFINE' => join(' ', @defines)) : ()),
    };
}

sub check_pkg_config {
    my ($pkg, $required_version) = @_;

    my $pc = 'pkg-config';
    my @errors = ();

    # Check that pkg-config is installed
    my $pcver = `$pc --version`;
    if ($pcver eq '') {
        push @errors, "$pc not found";
    }
    else {
        # Check that pkg-config knows about the package
        my $pkgexists = `$pc --exists $pkg`;
        if ($?) {
            push @errors, "Package $pkg not found by $pc";
        }
        else {
            # Check that the package is the right version
            my $pkgver = `$pc --atleast-version='$required_version' $pkg`;
            if ($?) {
                push @errors, "Package $pkg is not the right version (at least $required_version)";
            }
        }
    }

    return @errors;
}

sub pkg_config {
    my ($pkg, $option, $pc) = @_;

    $pc = 'pkg-config' unless defined $pc;

    my $str = `$pc --$option $pkg`;
    if ($?) {
        die "$pc --$option $pkg: $str\n";
    }
    else {
        chomp $str;
        return $str;
    }
}
